### DISCUSSION TITLE
disassembling intel hex file produces no results and indicates no errors

### DISCUSSION THREAD

**Discussion by wcolemanmoore** [radareorg/radare2#21742] [Q&A]
I'm running rasm2 to disassemble an intel hex file in Arm M0 32-bit processor format. When I run rasm2 it doesn't return any results or any error messages. It simply just does a CRLF and that is it.

Here is the command with arguments:

rasm2 -a arm -b 32 -f ihex://test.hex -O test.txt

In this case, an output file "test.txt" is created with a 0 byte length.

But no errors either!

I can't include the intel hex file I am using since it is for a customer.

Thanks for any assistance!

Coleman

![image](https://github.com/radareorg/radare2/assets/8653939/4591b5c0-99f3-470c-8be3-be5b9b2cffd6)

**Comment by trufae** [comment]
Rasm2 doesnt handle io plugins so the ihex:// thing shouldnt work. That should be an error

**Reply by trufae** [reply]
^ this is not correct. because rasm2 already handles ihex:// via -f. but you are missing the -d flag to instruct rasm2 to disassemble instead of the default operation which is assembling.

**Reply by trufae** [reply]
the reason why this line doesnt prodcues any error is probably because binary data is skipped/ignored during the assembler parsing, and therefor generates an empty file, which looks correct behaviour to me. But i also understand that it could be better not to ignore binary data when parsing text here

**Comment by trufae** [comment]
Seems like an useful usecase so please provide an ihex for testing and i'll find some time to support that in rasm2 meanwhile it should be possible to make a oneliner with r2 to do the same

**Reply by wcolemanmoore** [reply]
Ok, that explains the intel hex file input, but I also tried to read from a binary file (see screenshot), and got the same results, no file output and no errors.

**Reply by wcolemanmoore** [reply]
Please take a look at my rasm2 arguments so see why reading a binary file doesn't work. I'm sure I am missing something, but I don't know what it is.

![image](https://github.com/radareorg/radare2/assets/8653939/8f6ba0df-36d3-4632-80f1-644c04d40655)

Thanks.

**Reply by trufae** [reply]
This line works fine:
`$ rasm2 -a arm -b 32 -f file.bin -0 file.out`

the problem is assuming that the input is binary and that it will be disassembling, the primary purpose of `rasm2` is to assemble. if you want to disassemble you need to pass the `-d` flag. and the file.bin should be file.asm and contain plaintext assembly.

**Reply by trufae** [reply]
i can confirm that ihex was already supported via `-f `. so sorry for my previous answer because i was in the phone and couldnt check properly.

my file.bin contains "jeje", whichi s clearly not an ihex source, and therefor it shows an error

```
1$ rasm2 -a arm -b 32 -f ihex://file.bin -O file.out
ERROR: Invalid data in ihex file (jeje
)
ERROR: ihex: failed to parse file
ERROR: Cannot open file ihex://file.bin
1$
```

**Reply by wcolemanmoore** [reply]
Thank you for your comments and advice.

In all of the examples I have found concerning rasm2, the only disassembly I have seen is the -d flag operating on a string of hex-pairs. My customer has a 3100 line intel hex file that they want recovered back to asm and C if possible. Going through the records line by line with hex-pairs is difficult. Given that I have both an intel hex file and the equivalent binary file, is there a way to automate the recovery to asm and C? Whenever I include the -d flag on the command line, I receive an error about missing hex-pair data.

Thanks,

Coleman

**Reply by trufae** [reply]
Iirc the -B treats the input as binary
